pkg/kapis/events/v1alpha2/handler.go

```diff
@@ -0,0 +1,191 @@
+package v1alpha2
+
+import (
+	"errors"
+	"fmt"
+	"github.com/emicklei/go-restful"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apiserver/pkg/authentication/user"
+	"k8s.io/klog"
+	"kubesphere.io/kubesphere/pkg/api"
+	apiEvents "kubesphere.io/kubesphere/pkg/api/events/v1alpha2"
+	"kubesphere.io/kubesphere/pkg/apis/tenant/v1alpha1"
+	"kubesphere.io/kubesphere/pkg/apiserver/query"
+	"kubesphere.io/kubesphere/pkg/apiserver/request"
+	"kubesphere.io/kubesphere/pkg/constants"
+	"kubesphere.io/kubesphere/pkg/informers"
+	"kubesphere.io/kubesphere/pkg/models/events/v1alpha2"
+	"kubesphere.io/kubesphere/pkg/models/tenant"
+	"kubesphere.io/kubesphere/pkg/simple/client/events"
+	"kubesphere.io/kubesphere/pkg/utils/stringutils"
+	"strconv"
+	"strings"
+	"time"
+)
+
+type handler struct {
+	tenant tenant.Interface
+	eo     v1alpha2.EventsOperator
+}
+
+func newHandler(factory informers.InformerFactory, ec events.Client) *handler {
+	return &handler{tenant: tenant.New(factory), eo: v1alpha2.NewEventsOperator(ec)}
+}
+
+func (h handler) Query(req *restful.Request, resp *restful.Response) {
+	suser, ok := request.UserFrom(req.Request.Context())
+	if !ok {
+		err := errors.New("cannot obtain user info")
+		klog.Errorln(err)
+		api.HandleForbidden(resp, req, err)
+		return
+	}
+	userWorkspaces, userNamespaces, err := h.userScope(suser)
+	if err != nil {
+		klog.Errorln(err)
+		api.HandleInternalError(resp, req, err)
+		return
+	}
+	filter, err := h.parseFilterFrom(req, userWorkspaces, userNamespaces)
+	if err != nil {
+		klog.Errorln(err)
+		api.HandleInternalError(resp, req, err)
+		return
+	}
+
+	var ar apiEvents.APIResponse
+	if len(filter.InvolvedObjectNamespaceMap) == 0 { // does not query and returns empty result
+		switch req.QueryParameter("operation") {
+		case "histogram":
+			ar.Histogram = &events.Histogram{}
+		case "statistics":
+			ar.Statistics = &events.Statistics{}
+		default:
+			ar.Events = &events.Events{}
+		}
+		resp.WriteAsJson(ar)
+		return
+	}
+
+	switch req.QueryParameter("operation") {
+	case "histogram":
+		interval := req.QueryParameter("interval")
+		if interval == "" {
+			interval = "15m"
+		}
+		ar.Histogram, err = h.eo.CountOverTime(filter, interval)
+	case "statistics":
+		ar.Statistics, err = h.eo.StatisticsOnResources(filter)
+	default:
+		from, _ := strconv.ParseInt(req.QueryParameter("from"), 10, 64)
+		size, err := strconv.ParseInt(req.QueryParameter("size"), 10, 64)
+		if err != nil {
+			size = 10
+		}
+		sorts := map[string]string{
+			"lastTimestamp": req.QueryParameter("sort"),
```

Comment:
lastTimestamp in the sorts means?

Reply:
Like  the log query, It sorts the events